### ISSUE TITLE
Emit Metrics  about SDK internal state

### ISSUE THREAD

**Issue by cijothomas** [open-telemetry/opentelemetry-dotnet#3880]
SDK should report its internal statistics as metrics. Example include: # of instruments active, # of items processed by BatchExportProcessor, # of items dropped , # of metrics dropped with reason being a attribute/tag.

SDK can provide an extension method/option to let users enable the Meter.

**Comment by ohads-MSFT** [comment]
@cijothomas Any news on this by any chance? 
We are working on our monitoring experience and these would be very nice to have :)

**Comment by cijothomas** [comment]
No work has been done on this yet. @vishweshbankwar might have explored this a bit?

We also got feedback about adding this particularly for the Metrics being dropped when it hit maxstreamlimit, which is a common issue. Meanwhile, if there are any particular metrics you are interested in, please share them here, so we can take all feedback into account when working on this.

**Comment by cijothomas** [comment]
@dpk83 I believe you also mentioned the need to know how many metrics etc are being dropped, due to dimensioncap/other reasons.

**Comment by vishweshbankwar** [comment]
> No work has been done on this yet. @vishweshbankwar might have explored this a bit?
> 
> We also got feedback about adding this particularly for the Metrics being dropped when it hit maxstreamlimit, which is a common issue. Meanwhile, if there are any particular metrics you are interested in, please share them here, so we can take all feedback into account when working on this.

I have not started on this yet. I will be looking in to this soon.

**Comment by dpk83** [comment]
> @dpk83 I believe you also mentioned the need to know how many metrics etc are being dropped, due to dimensioncap/other reasons.

Yes, exactly. When metrics are dropped due to reaching the cardinality there needs to be some throttling signals that customers can use to get alerted and take appropriate action to fix the problems in time.

**Comment by ohads-MSFT** [comment]
BTW as a quick stopgap, you could just make a few properties in `BatchExportProcessor<T>` protected non-internal, and we could then subclass and fire our own custom metrics. Specifically `DroppedCount`, `ReceivedCount`, `ProcessedCount` would be good (current queue size would then be `ReceivedCount-DroppedCount-ProcessedCount`).

**Comment by cijothomas** [comment]
I think this is an important enough thing to be provided by the SDK itself for every user (i.e whoever is willing to pay for the cost of storing the additional metrics), without having to write extra code. Lets wait for the 1.4 ship to sail (~1-2 week), and get something in soon.

**Comment by kriskalish** [comment]
Is this still slated to go into the 1.5 release?

The proposed change is essential to understand if metrics/traces are giving limited visibility. Critical monitoring flows may rely on relatively complete visibility, so this change is excellent!

**Comment by cijothomas** [comment]
It got pushed out from 1.5, unfortunately. 
We should be doing this in 1.6. (it could be a different implementation for the metrics, to let users know if they are hitting Metric cardinality limits, because there is now a spec guiding how to deal with this.)
Please watch this issue for updates.

**Comment by cijothomas** [comment]
+1 to this issue from investigating issues with BatchExportProcessor which emits a [log about dropped items, but it does so only at the shutdown](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/BatchExportProcessor.cs#L220). This leaves us with no good way to know if BatchExportProcessor is dropping items due to buffer full, until we wait for the app to shutdown..

Maybe it can do the log in force_flush as well? So users can call a force flush occasionally to see if this is occurring. (as a temp. mitigation while the heartbeat/internalmetric feature gets added.)

**Comment by okonaraddi-msft** [comment]
> > @dpk83 I believe you also mentioned the need to know how many metrics etc are being dropped, due to dimensioncap/other reasons.
> 
> Yes, exactly. When metrics are dropped due to reaching the cardinality there needs to be some throttling signals that customers can use to get alerted and take appropriate action to fix the problems in time.

+1 on this. Some observability on the number of metric points would fantastic. My team is dealing with sporadic metric drops and we suspect it's due to cardinality limits being reached. We enabled the self-diagnostic logs recently, but we have to wait to see if there'll be any measurement drop logged. Having telemetry on the metric points would help us identify whether high cardinality is the issue and it would help us validate fix(es).

**Comment by cijothomas** [comment]
> My team is dealing with sporadic metric drops and we suspect it's due to cardinality limits being reached.

`OTEL_DOTNET_EXPERIMENTAL_METRICS_EMIT_OVERFLOW_ATTRIBUTE=true` is your best option right now to detect this. https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics#cardinality-limits

Internal metric about MetricPoint utilization will also help, but it is not coming in the immediate future, so the overflow attribute feature is a sure shot way of confirming if you are hit by cardinality limit or not!

**Comment by okonaraddi-msft** [comment]
@cijothomas Thank you for sharing! I'm having trouble following what's the behavior with overflow attribute enabled. We're using Histogram (from System.Diagnostics.Metrics) for our metrics so if we execute `someHistogram.Record(value, tags)` and there's an overflow, does that mean all metrics post-overflow will only have a tag called `otel.metric.overflow` set to `true` in lieu of the tags we provide?

**Comment by cijothomas** [comment]
> @cijothomas Thank you for sharing! I'm having trouble following what's the behavior with overflow attribute enabled. We're using Histogram (from System.Diagnostics.Metrics) for our metrics so if we execute `someHistogram.Record(value, tags)` and there's an overflow, does that mean all metrics post-overflow will only have a tag called `otel.metric.overflow` set to `true` in lieu of the tags we provide?

that is correct!

**Comment by djluck** [comment]
Just to further add that monitoring the number of processed + dropped items in the batch exporter would be critical for our services- without this, we'll have no visibility into if they are healthy or need to be tuned.

**Comment by ben-childs-docusign** [comment]
@TimothyMothra , @xiang17 any updates here? We are diagnosing missing spans in our telemetry pipelines and having metrics like dropped span count from the Batch Processor would be super helpful.

**Comment by cijothomas** [comment]
I'll work on getting this in soon. the semantic conventions exists now so can follow that. still need to be experimental, bout hopefully we'll stabilize semantic conventions alps very soon

**Comment by cijothomas** [comment]
@xiang17 so you mind if I take this one up?

**Comment by cijothomas** [comment]
https://github.com/open-telemetry/opentelemetry-dotnet/pull/7598
https://github.com/open-telemetry/opentelemetry-dotnet/pull/7486 added some metrics. More incoming.
